File: apps/server/src/jobs/steps/rollout.ts
import { appendFileSync, existsSync, readFileSync, rmSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { dirname, join } from 'node:path'
import process from 'node:process'
import { inspectContainer, listServiceContainers, removeContainer, stopContainer } from '@/docker/api'
import { setEnvLine } from '@/utils/env'

// Per-container health check timeout. Override via ROLLHOOK_HEALTH_TIMEOUT_MS env var.
const HEALTH_TIMEOUT_MS = Number(process.env.ROLLHOOK_HEALTH_TIMEOUT_MS) || 60_000
const POLL_INTERVAL_MS = 1_000

export async function rolloutApp(
  composePath: string,
  service: string,
  project: string,
  imageTag: string,
  logPath: string,
): Promise<void> {
  const log = (line: string) => appendFileSync(logPath, `[${new Date().toISOString()}] ${line}\n`)
  const cwd = dirname(composePath)

  // Merge IMAGE_TAG into a job-scoped temp env file — never touch the user's .env.
  // Docker Compose v2 .env takes precedence over shell env; --env-file with a temp
  // file overrides it without mutating the user's project directory.
  const userDotEnvPath = join(cwd, '.env')
  const userDotEnv = existsSync(userDotEnvPath) ? readFileSync(userDotEnvPath, 'utf8') : ''
  const mergedEnv = setEnvLine(userDotEnv, 'IMAGE_TAG', imageTag)
  const tmpEnvFile = join(tmpdir(), `rollhook-${crypto.randomUUID()}.env`)
  writeFileSync(tmpEnvFile, mergedEnv)

  try {
    // 1. Capture old container IDs before scale-up
    const oldContainers = await listServiceContainers(project, service)
    const oldIds = new Set(oldContainers.map(c => c.Id))
    // First deploy (no existing containers): start 1 replica, not 2
    const scaleCount = oldIds.size === 0 ? 1 : oldIds.size * 2

    log(`[rollout] Rolling out service: ${service} (IMAGE_TAG=${imageTag})`)
    log(`[rollout] Scaling service ${service} from ${oldIds.size}→${scaleCount} replicas`)

    // 2. Scale up via docker compose — the only remaining subprocess
    const scaleProc = Bun.spawn([
      'docker',
      'compose',
      '-f',
      composePath,
      '--env-file',
      tmpEnvFile,
      'up',
      '-d',
      '--no-recreate',
      '--scale',
      `${service}=${scaleCount}`,
    ], {
      cwd,
      stdout: 'pipe',
      stderr: 'pipe',
    })

    const [scaleExit, scaleStdout, scaleStderr] = await Promise.all([
      scaleProc.exited,
      new Response(scaleProc.stdout).text(),
      new Response(scaleProc.stderr).text(),
    ])

    if (scaleStdout.trim())
      log(`[rollout] ${scaleStdout.trim()}`)

    if (scaleExit !== 0)
      throw new Error(`docker compose up --scale failed (exit ${scaleExit}): ${scaleStderr}`)

    // 3. Identify new containers — poll briefly to account for Docker API propagation delay
    let newContainers = await findNewContainers(project, service, oldIds)
    if (newContainers.length === 0) {
      // Docker API may not reflect new containers immediately after compose returns
      for (let attempt = 0; attempt < 10; attempt++) {
        await new Promise(r => setTimeout(r, 500))
        newContainers = await findNewContainers(project, service, oldIds)
        if (newContainers.length > 0)
          break
      }
    }

    if (newContainers.length === 0)
      throw new Error(`Scale-up produced no new containers for service ${service} after 5s`)

    log(`[rollout] Waiting for ${newContainers.length} new container(s) to become healthy`)

    // 4. Wait for each new container to become healthy — each gets its own timeout
    for (let i = 0; i < newContainers.length; i++) {
      const container = newContainers[i]!
      log(`[rollout] Waiting for container ${container.Id.slice(0, 12)} to become healthy (${i + 1}/${newContainers.length})`)
      const containerStart = Date.now()
      const containerDeadline = containerStart + HEALTH_TIMEOUT_MS

      while (true) {
        if (Date.now() > containerDeadline) {
          await rollback(newContainers.map(c => c.Id), 'health check timed out', log)
          throw new Error(`Rolling deploy timed out: container ${container.Id.slice(0, 12)} did not become healthy within ${HEALTH_TIMEOUT_MS / 1000}s`)
        }

        let detail
        try {
          detail = await inspectContainer(container.Id)
        }
        catch (inspectErr) {
          await rollback(newContainers.map(c => c.Id), 'container inspection failed', log)
          throw inspectErr
        }
        const health = detail.State.Health

        if (health === null)
          throw new Error(`Container ${container.Id.slice(0, 12)} has no healthcheck — add a healthcheck to your compose service for zero-downtime deploys`)

        if (health.Status === 'healthy') {
          const elapsed = ((Date.now() - containerStart) / 1000).toFixed(1)
          log(`[rollout] Container ${container.Id.slice(0, 12)} healthy after ${elapsed}s`)
          break
        }

        if (health.Status === 'unhealthy') {
          await rollback(newContainers.map(c => c.Id), 'container became unhealthy', log)
          throw new Error(`Rolling deploy failed: container ${container.Id.slice(0, 12)} became unhealthy`)
        }

        await new Promise(r => setTimeout(r, POLL_INTERVAL_MS))
      }
    }

    // 5. All new containers healthy — drain old containers
    // Note: we remove old containers via the Docker API directly. Docker Compose discovers
    // current state from labels on next run, so this does not corrupt compose state.
    for (const oldId of oldIds) {
      log(`[rollout] Draining old container ${oldId.slice(0, 12)}`)
      await stopContainer(oldId)
      await removeContainer(oldId)
    }

    log(`[rollout] Service ${service} rolled out successfully`)
  }
  finally {
    try {
      rmSync(tmpEnvFile)
    }
    catch (e) {
      log(`[rollout] Warning: failed to remove temp env file ${tmpEnvFile}: ${e instanceof Error ? e.message : String(e)}`)
    }
  }
}

async function findNewContainers(
  project: string,
  service: string,
  oldIds: Set<string>,
) {
  const all = await listServiceContainers(project, service)
  return all.filter(c => !oldIds.has(c.Id))
}

async function rollback(newIds: string[], reason: string, log: (line: string) => void): Promise<void> {
  log(`[rollout] Rollback triggered: ${reason}`)
  for (const id of newIds) {
    log(`[rollout] Rollback: stopping new container ${id.slice(0, 12)}`)
    try {
      await stopContainer(id)
      await removeContainer(id)
    }
    catch (e) {
      log(`[rollout] Rollback cleanup error: ${e instanceof Error ? e.message : String(e)}`)
    }
  }
}

File: internal/jobs/steps/rollout.go
package steps

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"

	dockerpkg "github.com/jkrumm/rollhook/internal/docker"
)

const (
	defaultHealthTimeoutMS = 60_000
	pollIntervalMS         = 1_000
)

// Rollout performs a zero-downtime rolling deploy:
// scale up new containers, wait for health, drain old containers.
// On any failure it rolls back new containers before returning the error.
func Rollout(ctx context.Context, cli *client.Client, composePath, service, project, imageTag string, logFn func(string)) error {
	healthTimeoutMS := envInt("ROLLHOOK_HEALTH_TIMEOUT_MS", defaultHealthTimeoutMS)
	cwd := filepath.Dir(composePath)

	// 1. Capture old container IDs before scale-up.
	oldContainers, err := dockerpkg.ListServiceContainers(ctx, cli, project, service)
	if err != nil {
		return err
	}
	oldIDs := make(map[string]struct{}, len(oldContainers))
	for _, c := range oldContainers {
		oldIDs[c.ID] = struct{}{}
	}

	scaleCount := scaleTarget(len(oldIDs))
	logFn(fmt.Sprintf("[rollout] Rolling out service: %s (IMAGE_TAG=%s)", service, imageTag))
	logFn(fmt.Sprintf("[rollout] Scaling service %s from %d→%d replicas", service, len(oldIDs), scaleCount))

	// Write IMAGE_TAG into a job-scoped temp env file — never touch the user's .env.
	// Docker Compose v2 --env-file replaces the auto-loaded .env entirely, so we read
	// the user's .env (if present), merge IMAGE_TAG in, and write a temp file. This
	// preserves all the user's existing variables while overriding IMAGE_TAG.
	tmpEnv, err := writeTempEnv(cwd, imageTag)
	if err != nil {
		return fmt.Errorf("rollout: prepare env file: %w", err)
	}
	defer func() {
		if removeErr := os.Remove(tmpEnv); removeErr != nil {
			logFn(fmt.Sprintf("[rollout] Warning: failed to remove temp env file: %s", removeErr))
		}
	}()

	// 2. Scale up via docker compose subprocess.
	cmd := exec.CommandContext(ctx, "docker", "compose",
		"-f", composePath,
		"--env-file", tmpEnv,
		"up", "-d", "--no-recreate",
		"--scale", fmt.Sprintf("%s=%d", service, scaleCount),
	)
	cmd.Dir = cwd
	out, err := cmd.CombinedOutput()
	for _, line := range strings.Split(strings.TrimSpace(string(out)), "\n") {
		if line = strings.TrimSpace(line); line != "" {
			logFn(fmt.Sprintf("[rollout] %s", line))
		}
	}
	if err != nil {
		return fmt.Errorf("docker compose up --scale failed: %w", err)
	}

	// 3. Find new containers — poll to account for Docker API propagation delay.
	newContainers, err := pollNewContainers(ctx, cli, project, service, oldIDs)
	if err != nil {
		return err
	}
	newIDs := make([]string, len(newContainers))
	for i, c := range newContainers {
		newIDs[i] = c.ID
	}

	logFn(fmt.Sprintf("[rollout] Waiting for %d new container(s) to become healthy", len(newContainers)))

	// 4. Wait for each new container — each gets its own deadline.
	for i, c := range newContainers {
		id := c.ID
		logFn(fmt.Sprintf("[rollout] Waiting for container %s to become healthy (%d/%d)", id[:min(12, len(id))], i+1, len(newContainers)))

		if err := waitHealthy(ctx, cli, id, healthTimeoutMS, logFn); err != nil {
			rollbackContainers(cli, newIDs, err.Error(), logFn)
			return fmt.Errorf("rolling deploy failed: %w", err)
		}
	}

	// 5. All new containers healthy — drain old containers.
	for id := range oldIDs {
		short := id[:min(12, len(id))]
		logFn(fmt.Sprintf("[rollout] Draining old container %s", short))
		if err := dockerpkg.StopContainer(ctx, cli, id); err != nil {
			logFn(fmt.Sprintf("[rollout] Warning stopping %s: %s", short, err))
		}
		if err := dockerpkg.RemoveContainer(ctx, cli, id); err != nil {
			logFn(fmt.Sprintf("[rollout] Warning removing %s: %s", short, err))
		}
	}

	logFn(fmt.Sprintf("[rollout] Service %s rolled out successfully", service))
	return nil
}

// scaleTarget returns the number of replicas to scale up to.
// First deploy (no existing containers) starts 1 replica; subsequent deploys double.
func scaleTarget(currentCount int) int {
	if currentCount == 0 {
		return 1
	}
	return currentCount * 2
}

// waitHealthy polls container health until healthy, unhealthy, or timeout.
func waitHealthy(ctx context.Context, cli *client.Client, id string, timeoutMS int, logFn func(string)) error {
	short := id[:min(12, len(id))]
	deadline := time.Now().Add(time.Duration(timeoutMS) * time.Millisecond)
	start := time.Now()

	for {
		if time.Now().After(deadline) {
			return fmt.Errorf("container %s did not become healthy within %ds", short, timeoutMS/1000)
		}

		detail, err := dockerpkg.InspectContainer(ctx, cli, id)
		if err != nil {
			return fmt.Errorf("container inspection failed: %w", err)
		}

		if detail.State.Health == nil {
			return fmt.Errorf("container %s has no healthcheck — add a HEALTHCHECK to your service for zero-downtime deploys", short)
		}

		switch detail.State.Health.Status {
		case "healthy":
			elapsed := time.Since(start).Seconds()
			logFn(fmt.Sprintf("[rollout] Container %s healthy after %.1fs", short, elapsed))
			return nil
		case "unhealthy":
			return fmt.Errorf("container %s became unhealthy", short)
		}

		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(pollIntervalMS * time.Millisecond):
		}
	}
}

// rollbackContainers stops and removes the given container IDs.
// Uses a background context so cleanup completes even if the parent context is cancelled.
func rollbackContainers(cli *client.Client, ids []string, reason string, logFn func(string)) {
	bctx := context.Background()
	logFn(fmt.Sprintf("[rollout] Rollback triggered: %s", reason))
	for _, id := range ids {
		short := id[:min(12, len(id))]
		logFn(fmt.Sprintf("[rollout] Rollback: stopping new container %s", short))
		if err := dockerpkg.StopContainer(bctx, cli, id); err != nil {
			logFn(fmt.Sprintf("[rollout] Rollback cleanup error: %s", err))
		}
		if err := dockerpkg.RemoveContainer(bctx, cli, id); err != nil {
			logFn(fmt.Sprintf("[rollout] Rollback cleanup error: %s", err))
		}
	}
}

// pollNewContainers retries findNewContainers up to 10 times with 500ms delays
// to handle Docker API propagation delay after compose scale-up.
func pollNewContainers(ctx context.Context, cli *client.Client, project, service string, oldIDs map[string]struct{}) ([]container.Summary, error) {
	found, err := findNewContainers(ctx, cli, project, service, oldIDs)
	if err != nil {
		return nil, err
	}
	if len(found) > 0 {
		return found, nil
	}

	for attempt := 0; attempt < 10; attempt++ {
		time.Sleep(500 * time.Millisecond)
		found, err = findNewContainers(ctx, cli, project, service, oldIDs)
		if err != nil {
			return nil, err
		}
		if len(found) > 0 {
			return found, nil
		}
	}

	return nil, fmt.Errorf("scale-up produced no new containers for service %s after 5s", service)
}

func findNewContainers(ctx context.Context, cli *client.Client, project, service string, oldIDs map[string]struct{}) ([]container.Summary, error) {
	all, err := dockerpkg.ListServiceContainers(ctx, cli, project, service)
	if err != nil {
		return nil, err
	}
	var result []container.Summary
	for _, c := range all {
		if _, isOld := oldIDs[c.ID]; !isOld {
			result = append(result, c)
		}
	}
	return result, nil
}

// writeTempEnv reads the user's .env (if present), merges IMAGE_TAG, and writes
// a temp file. Returns the temp file path; caller is responsible for removing it.
func writeTempEnv(cwd, imageTag string) (string, error) {
	content := ""
	dotEnvPath := filepath.Join(cwd, ".env")
	if data, err := os.ReadFile(dotEnvPath); err == nil {
		content = string(data)
	} else if !os.IsNotExist(err) {
		return "", fmt.Errorf("reading .env: %w", err)
	}

	merged := setEnvLine(content, "IMAGE_TAG", imageTag)

	f, err := os.CreateTemp("", "rollhook-*.env")
	if err != nil {
		return "", err
	}
	defer f.Close()
	if _, err := f.WriteString(merged); err != nil {
		os.Remove(f.Name())
		return "", err
	}
	return f.Name(), nil
}

// setEnvLine merges KEY=VALUE into .env file content.
// Replaces the last occurrence of KEY= if present, appends otherwise.
// Preserves the file's trailing-newline convention.
func setEnvLine(content, key, value string) string {
	keyPrefix := key + "="
	newLine := key + "=" + value
	lines := strings.Split(content, "\n")

	lastIdx := -1
	for i, line := range lines {
		if strings.HasPrefix(line, keyPrefix) {
			lastIdx = i
		}
	}
	if lastIdx >= 0 {
		lines[lastIdx] = newLine
		return strings.Join(lines, "\n")
	}

	// Append, preserving the file's trailing-newline convention.
	if content == "" {
		return newLine
	}
	if strings.HasSuffix(content, "\n") {
		return content + newLine + "\n"
	}
	return content + "\n" + newLine
}

// envInt reads an integer env var, returning defaultVal if unset or unparseable.
func envInt(key string, defaultVal int) int {
	if v := os.Getenv(key); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			return n
		}
	}
	return defaultVal
}
